File: holo-client/src/main/java/com/alibaba/hologres/client/model/Column.java
/*
 * Copyright (c) 2022. Alibaba Group Holding Limited
 */

package com.alibaba.hologres.client.model;

import java.io.Serializable;
import java.sql.Types;
import java.util.Objects;

/** class for a Hologres column. */
public class Column implements Serializable {
    private String name;
    private String typeName;
    private int type;
    private String comment;
    private Boolean allowNull;
    private Boolean isPrimaryKey;
    private Object defaultValue;
    private Boolean isArrayType;
    private int arrayElementType;
    private int precision;
    private int scale;
    private Boolean isGeneratedColumn = false;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }

    public String getTypeName() {
        return typeName;
    }

    public void setTypeName(String typeName) {
        this.typeName = typeName;
    }

    public String getComment() {
        return comment;
    }

    public void setComment(String comment) {
        this.comment = comment;
    }

    public Boolean getAllowNull() {
        if (allowNull == null) {
            return false;
        }
        return allowNull;
    }

    public void setAllowNull(Boolean allowNull) {
        this.allowNull = allowNull;
    }

    public Boolean getPrimaryKey() {
        if (isPrimaryKey == null) {
            return false;
        }
        return isPrimaryKey;
    }

    public void setPrimaryKey(Boolean primaryKey) {
        isPrimaryKey = primaryKey;
    }

    public Object getDefaultValue() {
        return defaultValue;
    }

    public void setDefaultValue(Object defaultValue) {
        this.defaultValue = defaultValue;
    }

    public Boolean isArrayType() {
        return isArrayType;
    }

    public void setArrayType(Boolean arrayType) {
        this.isArrayType = arrayType;
    }

    public void setArrayElementType(String elementTypeName) {
        switch (elementTypeName) {
            case "char":
                this.arrayElementType = Types.CHAR;
                break;
            case "int2":
                this.arrayElementType = Types.SMALLINT;
                break;
            case "int4":
                this.arrayElementType = Types.INTEGER;
                break;
            case "int8":
                this.arrayElementType = Types.BIGINT;
                break;
            case "float4":
                this.arrayElementType = Types.REAL;
                break;
            case "float8":
                this.arrayElementType = Types.DOUBLE;
                break;
            case "bool":
                this.arrayElementType = Types.BIT;
                break;
            case "text":
            case "varchar":
                this.arrayElementType = Types.VARCHAR;
                break;
            default:
        }
    }

    public int getArrayElementType() {
        return arrayElementType;
    }

    public boolean isSerial() {
        return "serial".equals(typeName)
                || "bigserial".equals(typeName)
                || "smallserial".equals(typeName);
    }

    public int getPrecision() {
        return precision;
    }

    public void setPrecision(int precision) {
        this.precision = precision;
    }

    public int getScale() {
        return scale;
    }

    public void setScale(int scale) {
        this.scale = scale;
    }

    public void setGeneratedColumn(Boolean isGeneratedColumn) {
        this.isGeneratedColumn = isGeneratedColumn;
    }

    public Boolean isGeneratedColumn() {
        return isGeneratedColumn;
    }

    @Override
    public String toString() {
        return "\nColumn{"
                + "name='"
                + name
                + '\''
                + ", typeName='"
                + typeName
                + '\''
                + ((Objects.nonNull(allowNull) && !allowNull) ? ", not null" : "")
                + ((Objects.nonNull(defaultValue)) ? ", default value=" + defaultValue : "")
                + ((Objects.nonNull(isPrimaryKey) && isPrimaryKey) ? ", primary key" : "")
                + ((Objects.nonNull(isGeneratedColumn) && isGeneratedColumn)
                        ? ", generated column"
                        : "")
                + ((Objects.nonNull(comment) && !comment.isEmpty()) ? ", comment=" + comment : "")
                + '}';
    }
}

File: holo-client-py/hologres/column.py
from __future__ import annotations

from typing import Any, Optional

# SQL type constants (mirroring java.sql.Types used in the Java client)
BOOLEAN = 16
BIT = -7
TINYINT = -6
SMALLINT = 5
INTEGER = 4
BIGINT = -5
REAL = 7
DOUBLE = 8
NUMERIC = 2
DECIMAL = 3
CHAR = 1
VARCHAR = 12
DATE = 91
TIME = 92
TIMESTAMP = 93
TIMESTAMP_WITH_TIMEZONE = 2014
BINARY = -2
OTHER = 1111
ARRAY = 2003

# Map from PostgreSQL type name to SQL type constant
_PG_TYPE_MAP: dict[str, int] = {
    "bool": BOOLEAN,
    "boolean": BOOLEAN,
    "bit": BIT,
    "int2": SMALLINT,
    "smallint": SMALLINT,
    "smallserial": SMALLINT,
    "int4": INTEGER,
    "integer": INTEGER,
    "serial": INTEGER,
    "int8": BIGINT,
    "bigint": BIGINT,
    "bigserial": BIGINT,
    "float4": REAL,
    "real": REAL,
    "float8": DOUBLE,
    "double precision": DOUBLE,
    "numeric": NUMERIC,
    "decimal": DECIMAL,
    "money": NUMERIC,
    "char": CHAR,
    "character": CHAR,
    "bpchar": CHAR,
    "varchar": VARCHAR,
    "character varying": VARCHAR,
    "text": VARCHAR,
    "name": VARCHAR,
    "date": DATE,
    "time": TIME,
    "timetz": TIME,
    "time with time zone": TIME,
    "timestamp": TIMESTAMP,
    "timestamptz": TIMESTAMP_WITH_TIMEZONE,
    "timestamp with time zone": TIMESTAMP_WITH_TIMEZONE,
    "timestamp without time zone": TIMESTAMP,
    "bytea": BINARY,
    "json": VARCHAR,
    "jsonb": VARCHAR,
    "uuid": VARCHAR,
    "inet": VARCHAR,
    "cidr": VARCHAR,
    "macaddr": VARCHAR,
    "xml": VARCHAR,
    "point": OTHER,
    "line": OTHER,
    "polygon": OTHER,
    "circle": OTHER,
    "box": OTHER,
    "path": OTHER,
    "interval": OTHER,
    "roaringbitmap": BINARY,
}

# Array element type mapping
_ARRAY_ELEMENT_TYPE_MAP: dict[str, int] = {
    "char": CHAR,
    "int2": SMALLINT,
    "int4": INTEGER,
    "int8": BIGINT,
    "float4": REAL,
    "float8": DOUBLE,
    "bool": BIT,
    "boolean": BIT,
    "text": VARCHAR,
    "varchar": VARCHAR,
}


class Column:
    """Represents a single column definition in a Hologres table."""

    __slots__ = (
        "name",
        "type_name",
        "type",
        "comment",
        "allow_null",
        "is_primary_key",
        "default_value",
        "is_array_type",
        "array_element_type",
        "precision",
        "scale",
        "is_generated_column",
    )

    def __init__(
        self,
        name: str,
        type_name: str = "",
        type: int = OTHER,
        comment: Optional[str] = None,
        allow_null: bool = True,
        is_primary_key: bool = False,
        default_value: Any = None,
        is_array_type: bool = False,
        array_element_type: int = 0,
        precision: int = 0,
        scale: int = 0,
        is_generated_column: bool = False,
    ):
        self.name = name
        self.type_name = type_name
        self.type = type
        self.comment = comment
        self.allow_null = allow_null
        self.is_primary_key = is_primary_key
        self.default_value = default_value
        self.is_array_type = is_array_type
        self.array_element_type = array_element_type
        self.precision = precision
        self.scale = scale
        self.is_generated_column = is_generated_column

    @classmethod
    def from_pg_type_name(cls, name: str, pg_type_name: str, **kwargs) -> Column:
        """Create a Column from a PostgreSQL type name string."""
        raw = pg_type_name.strip().lower()
        is_array = raw.endswith("[]") or raw.startswith("_")

        if is_array:
            element_name = raw.rstrip("[]").lstrip("_")
            sql_type = ARRAY
            element_type = _ARRAY_ELEMENT_TYPE_MAP.get(element_name, VARCHAR)
            return cls(
                name=name,
                type_name=pg_type_name,
                type=sql_type,
                is_array_type=True,
                array_element_type=element_type,
                **kwargs,
            )

        sql_type = _PG_TYPE_MAP.get(raw, OTHER)
        return cls(name=name, type_name=pg_type_name, type=sql_type, **kwargs)

    @property
    def is_serial(self) -> bool:
        return self.type_name.lower() in ("serial", "bigserial", "smallserial")

    def __repr__(self) -> str:
        pk = " PK" if self.is_primary_key else ""
        null = "" if self.allow_null else " NOT NULL"
        return f"Column({self.name!r}, {self.type_name!r}{pk}{null})"
